modules/src/applications/ics20_fungible_token_transfer/context.rs

```diff
@@ -1,5 +1,326 @@
+use core::str::FromStr;
+
+use sha2::{Digest, Sha256};
+use subtle_encoding::hex;
+
+use super::error::Error as Ics20Error;
+use crate::applications::ics20_fungible_token_transfer::acknowledgement::Acknowledgement;
+use crate::applications::ics20_fungible_token_transfer::packet::PacketData;
+use crate::applications::ics20_fungible_token_transfer::relay_application_logic::on_ack_packet::process_ack_packet;
+use crate::applications::ics20_fungible_token_transfer::relay_application_logic::on_recv_packet::process_recv_packet;
+use crate::applications::ics20_fungible_token_transfer::{
+    DenomTrace, HashedDenom, IbcCoin, VERSION,
+};
+use crate::applications::ics20_fungible_token_transfer::events::{AckEvent, AckStatusEvent, RecvEvent, TimeoutEvent};
+use crate::applications::ics20_fungible_token_transfer::relay_application_logic::on_timeout_packet::process_timeout_packet;
+use crate::core::ics04_channel::channel::{Counterparty, Order};
 use crate::core::ics04_channel::context::{ChannelKeeper, ChannelReader};
+use crate::core::ics04_channel::msgs::acknowledgement::Acknowledgement as GenericAcknowledgement;
+use crate::core::ics04_channel::packet::Packet;
+use crate::core::ics04_channel::Version;
+use crate::core::ics05_port::context::{PortKeeper, PortReader};
+use crate::core::ics24_host::identifier::{ChannelId, ConnectionId, PortId};
+use crate::core::ics26_routing::context::{ModuleOutputBuilder, OnRecvPacketAck};
+use crate::prelude::*;
+use crate::signer::Signer;
+
+pub trait Ics20Keeper:
+    ChannelKeeper + PortKeeper + BankKeeper<AccountId = <Self as Ics20Keeper>::AccountId>
+{
+    type AccountId: Into<String>;
+
+    /// Sets a new {trace hash -> denom trace} pair to the store.
+    fn set_denom_trace(&mut self, denom_trace: &DenomTrace) -> Result<(), Ics20Error>;
+}
+
+pub trait Ics20Reader:
+    BankReader<AccountId = <Self as Ics20Reader>::AccountId>
+    + AccountReader<AccountId = <Self as Ics20Reader>::AccountId>
+    + ChannelReader
+    + PortReader
+{
+    type AccountId: Into<String> + FromStr;
+
+    /// get_port returns the portID for the transfer module.
+    fn get_port(&self) -> Result<PortId, Ics20Error>;
+
+    /// Returns the escrow account id for a port and channel combination
+    fn get_channel_escrow_address(
+        &self,
+        port_id: &PortId,
+        channel_id: ChannelId,
+    ) -> Result<<Self as Ics20Reader>::AccountId, Ics20Error> {
+        // https://github.com/cosmos/cosmos-sdk/blob/master/docs/architecture/adr-028-public-key-addresses.md
+        let contents = format!("{}/{}", port_id, channel_id);
+
+        let mut hasher = Sha256::new();
+        hasher.update(VERSION.as_bytes());
+        hasher.update(b"0");
+        hasher.update(contents.as_bytes());
+        let hash: Vec<u8> = hasher.finalize().to_vec().drain(20..).collect();
```

Comment:
The way `drain()` is used here, we're keeping everything *but the first 20 bytes*. That is, similar to `vec[20:]` in Python.

```suggestion
        let hasher_result = hasher.finalize();
        let hash: &[u8] = hasher_result.chunks(20).next().unwrap();
```

Reply:
Oops sorry, my bad. Nice catch btw! Fixed in ea0d62234ef387f27a8198128c0b1457f31d2bbf.